core/engine.go

```diff
@@ -106,128 +111,182 @@ func NewEngine(ex lib.ExecutionScheduler, o lib.Options, logger *logrus.Logger)
+//
+// Things to note:
+//  - The first lambda, Run(), synchronously executes the actual load test.
+//  - It can be prematurely aborted by cancelling the runCtx - this won't stop
+//    the metrics collection by the Engine.
+//  - Stopping the metrics collection can be done at any time after Run() has
+//    returned by cancelling the globalCtx
+//  - The second returned lambda can be used to wait for that process to finish.
+func (e *Engine) Init(globalCtx, runCtx context.Context) (run func() error, wait func(), err error) {
+	e.logger.Debug("Initialization starting...")
+	// TODO: if we ever need metrics processing in the init context, we can move
+	// this below the other components... or even start them concurrently?
+	if err := e.ExecutionScheduler.Init(runCtx, e.Samples); err != nil {
+		return nil, nil, err
 	}
-}
 
-func (e *Engine) Run(ctx context.Context) error {
-	e.runLock.Lock()
-	defer e.runLock.Unlock()
+	// TODO: move all of this in a separate struct? see main TODO above
+
+	runSubCtx, runSubCancel := context.WithCancel(runCtx)
+
+	resultCh := make(chan error)
+	runFn := func() error {
+		e.logger.Debug("Execution scheduler starting...")
+		err := e.ExecutionScheduler.Run(globalCtx, runSubCtx, e.Samples)
+		e.logger.WithError(err).Debug("Execution scheduler terminated")
+
+		select {
+		case <-runSubCtx.Done():
+			// do nothing, the test run was aborted somehow
+		default:
+			resultCh <- err // we finished normally, so send the result
+		}
 
-	e.logger.Debug("Engine: Starting with parameters...")
+		return err
+	}
+	waitFn := e.startBackgroundProcesses(globalCtx, runCtx, resultCh, runSubCancel)
+	return runFn, waitFn, nil
+}
 
-	collectorwg := sync.WaitGroup{}
-	collectorctx, collectorcancel := context.WithCancel(context.Background())
+// This starts a bunch of goroutines to process metrics, thresholds, and set the
+// test run status when it ends. It returns a function that can be used after
+// the provided context is called, to wait for the complete winding down of all
+// started goroutines.
+func (e *Engine) startBackgroundProcesses( //nolint:funlen
+	globalCtx, runCtx context.Context, runResult <-chan error, runSubCancel func(),
+) (wait func()) {
+	processes := new(sync.WaitGroup)
 
+	// Spin up all configured collectors
 	for _, collector := range e.Collectors {
-		collectorwg.Add(1)
+		processes.Add(1)
 		go func(collector lib.Collector) {
-			collector.Run(collectorctx)
-			collectorwg.Done()
+			collector.Run(globalCtx)
+			processes.Done()
 		}(collector)
 	}
 
-	subctx, subcancel := context.WithCancel(context.Background())
-	subwg := sync.WaitGroup{}
+	// Siphon and handle all produced metric samples
+	processes.Add(1)
+	go func() {
+		defer processes.Done()
+		e.processMetrics(globalCtx)
+	}()
 
-	// Run metrics emission.
-	subwg.Add(1)
+	// Run VU metrics emission, only while the test is running.
+	// TODO: move? this seems like something the ExecutionScheduler should emit...
```

Comment:
Possibly also looking at the implementation of runMetricsEmession I would argue it should put stuff in e.Samples and not call `ProcessSamples`

Reply:
Resolved by https://github.com/loadimpact/k6/pull/1390/commits/f2ca5b08beff01637aeba8d0e8d1ad27bf321374